File: 100-199/167_TwoSumIIInputArrayIsSorted.java
// Given an array of integers that is already sorted in ascending order, find two numbers such that they add up to a specific target number.

// The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.

// You may assume that each input would have exactly one solution.

// Input: numbers={2, 7, 11, 15}, target=9
// Output: index1=1, index2=2

//Idea: Binary Search for target - nums[i]. If found, return two indexes. If not, return null. The total time complexity, O(nlogn)

public class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int[] res = new int[2];
        for(int i = 0; i < numbers.length; i++){
            int temp = binarySearch(i + 1,numbers.length - 1,target - numbers[i],numbers);
            if(temp != -1){
                res[0] = i + 1;
                res[1] = temp + 1;
                return res;
            }
        }
        return res;
    }
    private int binarySearch(int low, int high, int target, int[] array){
        if(low > high) return -1;
        int mid = low + (high - low) / 2;
        if(array[mid] == target) return mid;
        if(array[mid] > target) return binarySearch(low, mid - 1, target, array);
        return binarySearch(mid + 1, high, target, array);
    }
}


// Solution 2: 
// Because numbers are sorted, use two pointers to traverse the array.
// i = 0, j = numbers.length - 1
// numbers[i] + numbers[j] < target, i++
// numbers[i] + numbers[j] > target, j--
// numbers[i] + numbers[j] == target, return.

public class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int[] res = new int[2];
        int i = 0, j = numbers.length - 1;
        while(i < j){
            int key = numbers[i] + numbers[j];
            if(key < target){
                i++;
            }else if(key > target){
                j--;
            }else{
                res[0] = i + 1;
                res[1] = j + 1;
                return res;
            }
        }
        return res;
    }
}

File: Javascript/100-199/167_Two_Sum_II_-_Input_array_is_sorted.js
//Given an array of integers that is already sorted in ascending order, find two numbers such that they add up to a specific target number.
//
//The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.
//
//You may assume that each input would have exactly one solution and you may not use the same element twice.
//
//
//Input: numbers={2, 7, 11, 15}, target=9
//Output: index1=1, index2=2
//

/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(numbers, target) {
    var res = [];
    for(let i = 0; i < numbers.length; i++){
        var index = binarySearch(numbers,target - numbers[i],i + 1,numbers.length - 1);
        if(index !== -1){
            res.push(i + 1);
            res.push(index + 1);
            return res;
        }
    }
    return res;
};

function binarySearch(array,key,start,end){
    if(end < start) return -1;
    var mid = Math.ceil(start + (end - start) / 2);
    if(array[mid] === key) return mid;
    else if(array[mid] < key) return binarySearch(array,key,mid + 1,end);
    return binarySearch(array,key,start,mid - 1);
}
